internal/provider/provider.go

```diff
@@ -233,13 +285,133 @@ func (p *pingfederateProvider) Configure(ctx context.Context, req provider.Confi
+	var clientSecret string
+	var scopes []string
+	var tokenUrl string
+
+	hasOAuthEnvVars := os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_ID") != "" || os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_SECRET") != "" || os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_TOKEN_URL") != ""
+
+	if internaltypes.IsNonEmptyObj(config.OAuth) || hasOAuthEnvVars {
+		configClientId, ok := config.OAuth.Attributes()["client_id"].(types.String)
+		if ok {
+			clientId = configClientId.ValueString()
+		} else {
+			clientId = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_ID")
+		}
+
+		configClientSecret, ok := config.OAuth.Attributes()["client_secret"].(types.String)
+		if ok {
+			clientSecret = configClientSecret.ValueString()
+		} else {
+			clientSecret = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_SECRET")
+		}
+
+		configScopes, ok := config.OAuth.Attributes()["scopes"].(types.List)
+		if ok {
+			configScopes.ElementsAs(ctx, &scopes, false)
+		} else {
+			scopes = []string{os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_SCOPES")}
+		}
+		if len(scopes) == 0 {
+			tflog.Info(ctx, "Unable to find scopes value to be used with OAuth authentication. If this is expected, this information can be ignored.")
+		}
+
+		configTokenUrl, ok := config.OAuth.Attributes()["token_url"].(types.String)
+		if ok {
+			tokenUrl = configTokenUrl.ValueString()
+		} else {
+			tokenUrl = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_TOKEN_URL")
+		}
+
+		if clientId != "" && clientSecret != "" && tokenUrl != "" {
+			hasOauthConfig = true
+		} else {
+			hasOauthConfig = false
+		}
+	}
+
+	// Validate the configuration
+	// User must provide a username and password, access token, or OAuth configuration to the provider
+	if !hasAccessTokenAuth && !hasBasicAuth && !hasOauthConfig {
+		resp.Diagnostics.AddError(
+			"Unable to find username and password, access_token, or OAuth authentication configuration",
+			"The username and password, access_token, or oauth configuration required values were not supplied. Either set them in the configuration or use the PINGFEDERATE_PROVIDER_* environment variables.",
+		)
+	} else if hasAccessTokenAuth && hasBasicAuth && hasOauthConfig {
+		resp.Diagnostics.AddError(
+			"Not all authentication method values can be used together",
+			"username and password, access_token, and oauth configuration cannot all be set. Only one of them can be set.",
+		)
+	} else {
+		// If user has not provided an OAuth configuration or access token, they must provide username and password
+		if !hasOauthConfig && !hasAccessTokenAuth && hasBasicAuth {
+			if username == "" {
+				returnAuthAttributeDiagsError("username", "basic", "PINGFEDERATE_PROVIDER_USERNAME", resp)
+			}
+
+			if password == "" {
+				returnAuthAttributeDiagsError("password", "basic", "PINGFEDERATE_PROVIDER_PASSWORD", resp)
+			}
+		}
+
+		// If user has not provided username and password or an OAuth configuration, they must provide an access token
+		if hasAccessTokenAuth && (!hasBasicAuth || !hasOauthConfig) {
```

Comment:
What if the user has set username, password, and an auth token? Looks like this will accept that and end up setting all three on the provider config.